hollow/src/main/java/com/netflix/hollow/core/memory/encoding/VarInt.java

```diff
@@ -167,6 +167,46 @@ public static int readVInt(ByteData arr, long position) {
         return value;
     }
 
+    /**
+     * Read a series of variable length integers from the supplied {@link ByteData} starting at the specified position.
+     * @param arr the byte data to read from
+     * @param position the position in the byte data to read from
+     * @param length the number of bytes that should be read
+     * @param output an array where outputs will be placed, which should be at least {@code length} long and zero.
+     * @return the number of values written
+     */
+    public static int readVIntsInto(ByteData arr, long position, int length, int[] output) {
+        int count = 0;
+        for(int i = 0; i < length; i++) {
+            int b = arr.get(position + i);
+
+            output[count] = (output[count] << 7) | (b & 0x7f);
+            count += (~b >> 7) & 0x1;
+        }
+
+        return count;
+    }
+
+    /**
+     * Read a series of variable length integers (as chars) from the supplied {@link ByteData} starting at the specified position.
+     * @param arr the byte data to read from
+     * @param position the position in the byte data to read from
+     * @param length the number of bytes that should be read
+     * @param output an array where outputs will be placed, which should be at least {@code length} long and zero.
+     * @return the number of values written
+     */
+    public static int readVIntsInto(ByteData arr, long position, int length, char[] output) {
+        int count = 0;
+        for(int i = 0; i < length; i++) {
+            int b = arr.get(position + i);
+
+            output[count] = (char) ((output[count] << 7) | (b & 0x7f));
```

Comment:
I don't know much about Hotspot but do you think [this optimization](https://github.com/protocolbuffers/protobuf/blob/main/java/core/src/main/java/com/google/protobuf/Utf8.java#L660-L669) (from protobuf decoding) could be something to try?

Reply:
Hard to explore with the error bars, however it seems that ascii-only strings gives a further notable improvement.

```
BEFORE FASTPATH

Benchmark                                               (countStrings)  (countStringsDb)  (maxStringLength)  (probabilityUnicode)  Mode  Cnt  Score   Error  Units
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                  5                     0  avgt   20  0.048 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                  5                    10  avgt   20  0.049 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 25                     0  avgt   20  0.059 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 25                    10  avgt   20  0.059 ± 0.003  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 50                     0  avgt   20  0.080 ± 0.003  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 50                    10  avgt   20  0.076 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                150                     0  avgt   20  0.165 ± 0.006  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                150                    10  avgt   20  0.179 ± 0.009  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000               1000                     0  avgt   20  0.907 ± 0.040  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000               1000                    10  avgt   20  0.921 ± 0.043  ms/op


AFTER FASTPATH

Benchmark                                               (countStrings)  (countStringsDb)  (maxStringLength)  (probabilityUnicode)  Mode  Cnt  Score   Error  Units
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                  5                     0  avgt   20  0.045 ± 0.004  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                  5                    10  avgt   20  0.044 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 25                     0  avgt   20  0.049 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 25                    10  avgt   20  0.047 ± 0.002  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 50                     0  avgt   20  0.068 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                 50                    10  avgt   20  0.072 ± 0.001  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                150                     0  avgt   20  0.157 ± 0.003  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000                150                    10  avgt   20  0.171 ± 0.007  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000               1000                     0  avgt   20  0.602 ± 0.048  ms/op
HollowObjectTypeReadStateShardBenchmark.testReadString             500            100000               1000                    10  avgt   20  0.938 ± 0.027  ms/op
```

Reply:
Oh, wow